File: wakakusa/tools/gen.rb
#!/usr/bin/env ruby
# frozen_string_literal: true
#
# Read elements.toml and write the three files that have to agree about
# it: the numbers both sides count with, the Ruby an app calls, and the
# Rust constants the engine reads. Nothing here is written by hand, so
# an element cannot mean one thing in Ruby and another in the engine.
#
#   tools/gen.rb            write the files
#   tools/gen.rb --check    fail if what is on disk is not what this
#                           would write (the sweep runs this)
#
# The TOML this reads is the TOML we write: array-of-tables headers,
# scalars, and arrays of one-line inline tables. The Rust side parses
# the same file with a real parser, and its test fails if a key here
# never reached an arm — so a mistake in the reader below cannot pass
# quietly.

require "fileutils"

ROOT = File.expand_path("..", __dir__)
REPO = File.expand_path("..", ROOT)

# --- the TOML this file speaks ----------------------------------------------

def toml_value(s)
  case s
  when /\A"(.*)"\z/m then Regexp.last_match(1).gsub('\\"', '"').gsub("\\\\", "\\")
  when "true" then true
  when "false" then false
  when "[]" then []
  when /\A-?\d+\z/ then s.to_i
  else s.to_f
  end
end

def toml_inline(s)
  h = {}
  s.scan(/(\w+)\s*=\s*("(?:[^"\\]|\\.)*"|\[\]|[^,}\s]+)/) { |k, v| h[k] = toml_value(v) }
  h
end

def parse_toml(text)
  doc = {}
  cur = nil
  key = nil
  arr = nil
  text.each_line do |raw|
    line = raw.strip
    next if line.empty? || line.start_with?("#")
    if arr
      if line.start_with?("]")
        cur[key] = arr
        arr = nil
      elsif line.start_with?("{")
        arr << toml_inline(line)
      end
      next
    end
    if (m = line.match(/\A\[\[(\w+)\]\]\z/))
      cur = {}
      (doc[m[1]] ||= []) << cur
      next
    end
    k, v = line.split("=", 2)
    k = k.strip
    v = v.strip
    if v == "["
      key = k
      arr = []
    else
      cur[k] = toml_value(v)
    end
  end
  doc
end

TABLE = parse_toml(File.read(File.join(ROOT, "elements.toml")))
RIDERS = TABLE.fetch("rider")
ELEMENTS = TABLE.fetch("element")

# --- the numbers ------------------------------------------------------------

# A keyword is numbered once, by its name, however many elements take
# it: `width` is one key whether a button owns it or a box rides on it.
KEY_ID = {}
def key_id(name)
  KEY_ID[name] ||= KEY_ID.length + 1
end
RIDERS.each { |r| key_id(r.fetch("name")) }
ELEMENTS.each { |e| e.fetch("props").each { |p| key_id(p.fetch("name")) } }

KIND_ID = {}
ELEMENTS.each_with_index { |e, i| KIND_ID[e.fetch("name")] = i + 1 }

PAYLOADS = { "none" => 0, "text" => 1, "bool" => 2, "int" => 3, "float" => 4 }.freeze

def const(name) = name.upcase

def handler?(prop) = prop.key?("handler")

# What a prop's Ruby default reads as at the call site.
def ruby_default(prop)
  return "nil" if handler?(prop)
  d = prop["default"]
  case prop.fetch("type")
  when "str" then d.inspect
  when "num" then format("%.1f", d)
  when "int" then d.to_s
  when "bool" then d.to_s
  when "strs", "nums", "nums2" then "[]"
  when "rows" then "nil"
  end
end

# A rider whose box is built because somebody WROTE it arrives as nil.
def rider_default(rider) = rider["presence"] ? "nil" : nil

def banner(tool)
  "# Generated from elements.toml by tools/gen.rb. Do not edit by hand;\n" \
    "# edit the table and run `tools/gen.rb`#{tool}.\n"
end

# --- lib/wakakusa/keys.rb ---------------------------------------------------

def gen_keys
  out = +""
  out << banner("")
  out << "\n"
  out << "# The numbers the two sides of the engine's C face count with.\n"
  out << "module WK\n"
  out << "  # What a handler is called with.\n"
  PAYLOADS.each { |name, n| out << "  PAY_#{const(name)} = #{n}\n" }
  out << "\n  # The elements.\n"
  KIND_ID.each { |name, n| out << "  KIND_#{const(name)} = #{n}\n" }
  out << "\n  # The keyword arguments, numbered once across every element.\n"
  KEY_ID.each { |name, n| out << "  K_#{const(name)} = #{n}\n" }
  out << "end\n"
  out
end

# --- lib/wakakusa/elements.rb -----------------------------------------------

def send_call(prop, var)
  key = "WK::K_#{const(prop.fetch("name"))}"
  case prop.fetch("type")
  when "str" then "PixieC.pixie_str(el, #{key}, #{var})"
  when "num" then "PixieC.pixie_num(el, #{key}, #{var})"
  when "int" then "PixieC.pixie_int(el, #{key}, #{var})"
  when "bool" then "PixieC.pixie_bool(el, #{key}, #{var} ? 1 : 0)"
  end
end

def prop_lines(prop)
  name = prop.fetch("name")
  key = "WK::K_#{const(name)}"
  return ["  wakakusa_send_#{prop.fetch("handler")}(el, #{key}, #{name}) unless #{name}.nil?"] if handler?(prop)

  case prop.fetch("type")
  when "strs"
    ["  #{name}.each { |v| PixieC.pixie_push_str(el, #{key}, v) }"]
  when "nums"
    ["  #{name}.each { |v| PixieC.pixie_push_num(el, #{key}, v) }"]
  when "nums2"
    ["  #{name}.each do |inner|",
     "    PixieC.pixie_list_break(el, #{key})",
     "    inner.each { |v| PixieC.pixie_push_num(el, #{key}, v) }",
     "  end"]
  when "rows"
    ["  wakakusa_rows(el, #{key}, &blk)"]
  else
    call = send_call(prop, name)
    # A positional argument is always written, so it always crosses; a
    # keyword only crosses when it differs from what the engine would
    # have used anyway.
    prop["pos"] ? ["  #{call}"] : ["  #{call} if #{name} != #{ruby_default(prop)}"]
  end
end

def signature(el)
  props = el.fetch("props")
  head = []
  head << "*kids" if el["children"]
  props.select { |p| p["pos"] }.reject { |p| p["type"] == "rows" }.each do |p|
    head << (p.key?("default") ? "#{p.fetch("name")} = #{ruby_default(p)}" : p.fetch("name"))
  end
  props.reject { |p| p["pos"] }.each { |p| head << "#{p.fetch("name")}: #{ruby_default(p)}" }
  head << "**riders"
  head << "&blk" if el["primary"] || props.any? { |p| p["type"] == "rows" }
  head.join(", ")
end

def wrap_signature(name, sig)
  line = "def #{name}(#{sig})"
  return [line] if line.length <= 78

  parts = sig.split(", ")
  lines = []
  cur = +"def #{name}("
  indent = " " * cur.length
  parts.each_with_index do |p, i|
    piece = p + (i == parts.length - 1 ? "" : ",")
    if cur.length + piece.length > 78 && cur.strip != "def #{name}("
      lines << cur.rstrip
      cur = indent.dup
    end
    cur << piece << " "
  end
  lines << "#{cur.rstrip})"
  lines
end

def gen_elements
  out = +""
  out << banner("")
  out << "\n"
  out << "# One method per element. Its own keywords are spelled out; the\n"
  out << "# keywords every element takes ride in `riders` and are applied\n"
  out << "# below, so an element's own `width` wins over the box's.\n"
  ELEMENTS.each do |el|
    name = el.fetch("name")
    out << "\n"
    doc = el["doc"]
    if doc
      words = doc.split(" ")
      line = +"#"
      words.each do |w|
        if line.length + w.length + 1 > 74
          out << line << "\n"
          line = +"#"
        end
        line << " " << w
      end
      out << line << "\n"
    end
    out << wrap_signature(name, signature(el)).join("\n") << "\n"
    out << "  el = PixieC.pixie_el(WK::KIND_#{const(name)})\n"
    el.fetch("props").each do |p|
      if el["primary"] == p["name"]
        # The block is the usual way to write this element's handler; a
        # symbol naming one of the app's methods is the other. Only a
        # literal block survives being handed on in a compiled run, so
        # the two paths stay apart here rather than meeting in a local.
        k = "WK::K_#{const(p.fetch("name"))}"
        h = p.fetch("handler")
        out << "  if #{p.fetch("name")}.nil?\n"
        out << "    wakakusa_on_#{h}(el, #{k}, &blk) unless blk.nil?\n"
        out << "  else\n"
        out << "    wakakusa_send_#{h}(el, #{k}, #{p.fetch("name")})\n"
        out << "  end\n"
      else
        prop_lines(p).each { |l| out << l << "\n" }
      end
    end
    out << "  wakakusa_children(el, kids)\n" if el["children"]
    out << "  wakakusa_riders(el, riders, #{!!el["owns_label"]})\n"
    out << "  PixieC.pixie_end(el)\n"
    out << "end\n"
  end

  out << "\n"
  out << "# The keywords every element takes. An element that owns one of\n"
  out << "# these names under its own meaning never gets here: Ruby binds it\n"
  out << "# to that element's own keyword first.\n"
  out << "def wakakusa_rider(el, name, v)\n"
  out << "  case name\n"
  RIDERS.each do |r|
    key = "WK::K_#{const(r.fetch("name"))}"
    call = case r.fetch("type")
           when "str" then "PixieC.pixie_str(el, #{key}, v)"
           when "num" then "PixieC.pixie_num(el, #{key}, v)"
           when "int" then "PixieC.pixie_int(el, #{key}, v)"
           when "bool" then "PixieC.pixie_bool(el, #{key}, v ? 1 : 0)"
           end
    out << "  when :#{r.fetch("name")} then #{call}\n"
  end
  out << "  else\n"
  out << "    raise ArgumentError, \"no property `\#{name}` — every element takes \" \\\n"
  out << "                         \"#{RIDERS.map { |r| r.fetch("name") }.join(", ")}\"\n"
  out << "  end\n"
  out << "end\n"
  out
end


# --- lib/wakakusa/table.rb ---------------------------------------------------

def gen_table
  out = +""
  out << banner("")
  out << "\n"
  out << "# What the checker reads: which keywords each element takes, and\n"
  out << "# which of them name a handler.\n"
  out << "module WK\n"
  out << "  RIDERS = %w[#{RIDERS.map { |r| r.fetch("name") }.join(" ")}].freeze\n"
  out << "\n  ELEMENTS = {\n"
  ELEMENTS.each do |el|
    own = el.fetch("props").map { |p| p.fetch("name") }
    own -= ["row"] if el.fetch("props").any? { |p| p["type"] == "rows" }
    allowed = own + RIDERS.map { |r| r.fetch("name") } - (el["owns_label"] ? ["a11y_label"] : [])
    out << "    \"#{el.fetch("name")}\" => %w[#{allowed.uniq.sort.join(" ")}],\n"
  end
  out << "  }.freeze\n"
  handlers = ELEMENTS.flat_map { |el| el.fetch("props").select { |p| handler?(p) }.map { |p| p.fetch("name") } }
  out << "\n  HANDLERS = %w[#{handlers.uniq.sort.join(" ")}].freeze\n"
  out << "end\n"
  out
end

# --- crates/pixie-capi/src/vocab.rs -------------------------------------------

def rust_str(s) = s.inspect

def gen_rust
  out = +""
  out << "//! Generated from `wakakusa/elements.toml` by `wakakusa/tools/gen.rb`.\n"
  out << "//! Do not edit by hand; edit the table and run the generator.\n"
  out << "//!\n"
  out << "//! The numbers here are the ones the caller's side counts with, and\n"
  out << "//! the defaults are what a property reads as when the caller left it\n"
  out << "//! out — which is why only what an app actually wrote has to cross.\n"
  out << "\n#![allow(dead_code)]\n\n"
  out << "/// What a handler is called with. The engine names the kind when\n"
  out << "/// it hands an event over, because a caller keeps one registry per\n"
  out << "/// kind: a list of blocks all called with the same sort of value is\n"
  out << "/// one a compiler can type, and a mixed one is not.\n"
  PAYLOADS.each { |name, n| out << "pub const PAY_#{const(name)}: i64 = #{n};\n" }
  out << "\n"
  KIND_ID.each { |name, n| out << "pub const KIND_#{const(name)}: i32 = #{n};\n" }
  out << "\n"
  KEY_ID.each { |name, n| out << "pub const K_#{const(name)}: i32 = #{n};\n" }

  defs = { "num" => [], "int" => [], "bool" => [], "str" => [] }
  ELEMENTS.each do |el|
    kind = "KIND_#{const(el.fetch("name"))}"
    el.fetch("props").each do |p|
      next if handler?(p) || p["pos"]
      key = "K_#{const(p.fetch("name"))}"
      case p.fetch("type")
      when "num" then defs["num"] << "    (#{kind}, #{key}, #{format("%.1f", p["default"])}),"
      when "int" then defs["int"] << "    (#{kind}, #{key}, #{p["default"]}),"
      when "bool" then defs["bool"] << "    (#{kind}, #{key}, #{p["default"]}),"
      when "str" then defs["str"] << "    (#{kind}, #{key}, #{rust_str(p["default"])}),"
      end
    end
  end
  RIDERS.each do |r|
    next if r["presence"]
    key = "K_#{const(r.fetch("name"))}"
    case r.fetch("type")
    when "num" then defs["num"] << "    (0, #{key}, #{format("%.1f", r["default"])}),"
    when "int" then defs["int"] << "    (0, #{key}, #{r["default"]}),"
    when "bool" then defs["bool"] << "    (0, #{key}, #{r["default"]}),"
    when "str" then defs["str"] << "    (0, #{key}, #{rust_str(r["default"])}),"
    end
  end
  out << "\n/// `(kind, key, value)` — what a property reads as when nobody wrote\n"
  out << "/// it. Kind 0 is a rider, which means the same on every element.\n"
  { "num" => "f64", "int" => "i64", "bool" => "bool", "str" => "&str" }.each do |t, ty|
    out << "pub const DEF_#{t.upcase}: &[(i32, i32, #{ty})] = &[\n"
    out << defs[t].join("\n") << "\n];\n"
  end

  out << "\n/// Every element and every keyword the table declares, by name, so\n"
  out << "/// a test can read the table and check that each one reached an arm.\n"
  out << "pub const KINDS: &[(&str, i32)] = &[\n"
  KIND_ID.each_key { |name| out << "    (#{rust_str(name)}, KIND_#{const(name)}),\n" }
  out << "];\n"
  out << "pub const KEYS: &[(&str, i32)] = &[\n"
  KEY_ID.each_key { |name| out << "    (#{rust_str(name)}, K_#{const(name)}),\n" }
  out << "];\n"

  %w[size width height].each do |flavor|
    kinds = ELEMENTS.select { |e| e["native"] == flavor }.map { |e| "KIND_#{const(e.fetch("name"))}" }
    out << "\n/// The elements that size their own #{flavor == "size" ? "width and height" : flavor},\n"
    out << "/// so the box rider leaves that side alone.\n"
    out << "pub const NATIVE_#{flavor.upcase}: &[i32] = &[#{kinds.join(", ")}];\n"
  end
  out
end

# --- write, or check --------------------------------------------------------

FILES = {
  File.join(ROOT, "lib", "wakakusa", "keys.rb") => gen_keys,
  File.join(ROOT, "lib", "wakakusa", "elements.rb") => gen_elements,
  File.join(ROOT, "lib", "wakakusa", "table.rb") => gen_table,
  File.join(REPO, "crates", "pixie-capi", "src", "vocab.rs") => gen_rust,
}.freeze

check = ARGV.include?("--check")
stale = []
FILES.each do |path, body|
  if check
    stale << path unless File.exist?(path) && File.read(path) == body
  else
    FileUtils.mkdir_p(File.dirname(path))
    File.write(path, body)
    puts "wrote #{path.delete_prefix("#{REPO}/")}"
  end
end

if check && !stale.empty?
  warn "gen: these are not what elements.toml says — run wakakusa/tools/gen.rb"
  stale.each { |p| warn "  #{p.delete_prefix("#{REPO}/")}" }
  exit 1
end

File: rakugan/tools/gen.pl
#!/usr/bin/env perl
# Read the engine's element table (crates/pixie-capi/elements.toml) and
# write the two Perl files that have to agree about it: the table as
# Perl data, which the door's runtime and the translator both read, and
# one sub per element for an app to call. Nothing here is written by
# hand, so an element cannot mean one thing to the interpreted run and
# another to the compiled one.
#
#   tools/gen.pl            write the files
#   tools/gen.pl --check    fail if what is on disk is not what this
#                           would write (the sweep runs this)
#
# The TOML this reads is the TOML we write: array-of-tables headers,
# scalars, and arrays of one-line inline tables — the same subset
# wakakusa/tools/gen.rb speaks. The Rust test beside the table parses
# it with a real parser and fails if a key never reached an arm, so a
# mistake in the reader below cannot pass quietly.
use strict;
use warnings;
use utf8;
use open qw(:std :encoding(UTF-8));
use FindBin;
use File::Spec;

my $ROOT = File::Spec->rel2abs(File::Spec->catdir($FindBin::Bin, '..'));   # rakugan/
my $REPO = File::Spec->rel2abs(File::Spec->catdir($ROOT, '..'));
my $TABLE_PATH = "$REPO/crates/pixie-capi/elements.toml";

# --- the TOML this file speaks ----------------------------------------------

sub toml_value {
    my ($s) = @_;
    return $1 =~ s/\\"/"/gr =~ s/\\\\/\\/gr if $s =~ /\A"(.*)"\z/s;
    return 1 if $s eq 'true';
    return 0 if $s eq 'false';
    return [] if $s eq '[]';
    return $s + 0 if $s =~ /\A-?\d+\z/;
    return $s + 0.0;
}

sub toml_inline {
    my ($s) = @_;
    my %h;
    while ($s =~ /(\w+)\s*=\s*("(?:[^"\\]|\\.)*"|\[\]|[^,}\s]+)/g) {
        $h{$1} = toml_value($2);
    }
    # A boolean has to stay apart from a number when it is written back.
    $h{"_bool_$1"} = 1 while $s =~ /(\w+)\s*=\s*(?:true|false)\b/g;
    return \%h;
}

sub parse_toml {
    my ($text) = @_;
    my (%doc, $cur, $key, $arr);
    for my $raw (split /\n/, $text) {
        (my $line = $raw) =~ s/\A\s+|\s+\z//g;
        next if $line eq '' || $line =~ /\A#/;
        if ($arr) {
            if ($line =~ /\A\]/) { $cur->{$key} = $arr; undef $arr }
            elsif ($line =~ /\A\{/) { push @$arr, toml_inline($line) }
            next;
        }
        if ($line =~ /\A\[\[(\w+)\]\]\z/) {
            $cur = {};
            push @{ $doc{$1} }, $cur;
            next;
        }
        my ($k, $v) = split /=/, $line, 2;
        $k =~ s/\s+\z//;
        $v =~ s/\A\s+//;
        if ($v eq '[') { $key = $k; $arr = [] }
        else {
            $cur->{$k} = toml_value($v);
            $cur->{"_bool_$k"} = 1 if $v =~ /\A(?:true|false)\z/;
        }
    }
    return \%doc;
}

open my $tfh, '<:encoding(UTF-8)', $TABLE_PATH or die "$TABLE_PATH: $!\n";
my $TABLE = parse_toml(do { local $/; <$tfh> });
close $tfh;
my @RIDERS = @{ $TABLE->{rider} };
my @ELEMENTS = @{ $TABLE->{element} };
my @OPS = @{ $TABLE->{op} };

# A drawing command's values, in the order the C face takes them: the
# name an app writes, the name the `.pix` writes, and the type.
for my $op (@OPS) {
    $op->{params} = [map {
        my ($lhs, $rhs) = split /:\s*/, $_, 2;
        my ($name, $pix) = split /=/, $lhs, 2;
        my ($ty, $default) = split /\s*=\s*/, $rhs, 2;
        { name => $name, pix => $pix // $name, type => $ty,
          (defined $default ? (default => $default) : ()) }
    } split /,\s*/, $op->{args}];
}

# --- the numbers ------------------------------------------------------------
# A keyword is numbered once, by its name, however many elements take
# it — the same numbering gen.rb gives, since both read one table in
# one order.
my (%KEY_ID, @KEY_ORDER);
sub key_id {
    my ($name) = @_;
    return $KEY_ID{$name} //= do { push @KEY_ORDER, $name; scalar @KEY_ORDER };
}
key_id($_->{name}) for @RIDERS;
for my $e (@ELEMENTS) { key_id($_->{name}) for @{ $e->{props} } }
my %KIND_ID;
$KIND_ID{ $ELEMENTS[$_]{name} } = $_ + 1 for 0 .. $#ELEMENTS;

my %PAYLOADS = (none => 0, text => 1, bool => 2, int => 3, float => 4);
my @PAYLOAD_ORDER = qw(none text bool int float);
# What a handler of each kind is called with: the .pix type of the
# parameter, and what the payload is called at a .pix call site.
my %PAYLOAD_TYPE = (text => 'String', bool => 'Bool', int => 'Int', float => 'Float');
my %PAYLOAD_NAME = (none => '', text => 'text', bool => 'checked', int => 'index', float => 'value');

# --- the .pix spelling ------------------------------------------------------

sub camel { my ($s) = @_; $s =~ s/_(\w)/\u$1/g; return $s }
sub pix_element { my ($el) = @_; return $el->{pix} // ucfirst camel($el->{name}) }
sub pix_prop { my ($p) = @_; return $p->{pix} // camel($p->{name}) }

# --- Perl text --------------------------------------------------------------

sub q_str { my ($s) = @_; $s =~ s/([\\'])/\\$1/g; return "'$s'" }

sub perl_default {
    my ($p) = @_;
    return 'undef' unless exists $p->{default};
    my $d = $p->{default};
    my $t = $p->{type};
    return q_str($d) if $t eq 'str';
    return ($d ? 1 : 0) if $t eq 'bool';
    return sprintf('%.1f', $d) if $t eq 'num';
    return "$d" if $t eq 'int';
    return '[]';
}

sub banner {
    return "# Generated from crates/pixie-capi/elements.toml by tools/gen.pl. Do not\n"
         . "# edit by hand; edit the table and run `tools/gen.pl`.\n";
}

# --- lib/Rakugan/Vocab.pm ---------------------------------------------------
# The table as Perl data. Read by the door's runtime under the app's
# perl and by the translator under the tooling's, so it is plain data
# and nothing newer than perl 5.34 is written here.

sub gen_vocab {
    my $out = "package Rakugan::Vocab;\n" . banner() . <<'HEAD';
#
# The engine's vocabulary as data: the numbers both sides of the C face
# count with, every element with its keywords, and the keywords every
# element takes. The door's runtime writes an element from this; the
# translator reads the same rows to write the .pix.
use strict;
use warnings;
use utf8;

HEAD
    $out .= "# What a handler is called with.\n";
    $out .= "our %PAY = (" . join(', ', map { "$_ => $PAYLOADS{$_}" } @PAYLOAD_ORDER) . ");\n";
    $out .= "# The .pix type of a handler's parameter, and the payload's name at a .pix call site.\n";
    $out .= "our %PAYLOAD_TYPE = (" . join(', ', map { "$_ => '$PAYLOAD_TYPE{$_}'" } qw(text bool int float)) . ");\n";
    $out .= "our %PAYLOAD_NAME = (" . join(', ', map { "$_ => '$PAYLOAD_NAME{$_}'" } qw(text bool int float)) . ");\n\n";
    $out .= "# The elements.\n";
    $out .= "our %KIND = (\n" . join('', map { "    $_->{name} => $KIND_ID{$_->{name}},\n" } @ELEMENTS) . ");\n\n";
    $out .= "# The keyword arguments, numbered once across every element.\n";
    $out .= "our %KEY = (\n" . join('', map { "    $_ => $KEY_ID{$_},\n" } @KEY_ORDER) . ");\n\n";

    $out .= "# The keywords every element takes. `presence` marks the four whose box\n";
    $out .= "# is built because somebody wrote the property; `owned` names the element\n";
    $out .= "# props that take the same word first.\n";
    $out .= "our \@RIDERS = (\n";
    for my $r (@RIDERS) {
        my @f = ("name => " . q_str($r->{name}), "type => " . q_str($r->{type}), "key => $KEY_ID{$r->{name}}",
                 "pix => " . q_str($r->{pix} // camel($r->{name})));
        push @f, "presence => 1" if $r->{presence};
        push @f, "default => " . perl_default($r) if exists $r->{default};
        push @f, "owned => " . q_str($r->{owned}) if defined $r->{owned};
        $out .= "    { " . join(', ', @f) . " },\n";
    }
    $out .= ");\nour %RIDER = map { \$_->{name} => \$_ } \@RIDERS;\n\n";

    $out .= "# The elements, in the table's order: the kind, the .pix element, what\n";
    $out .= "# it is (a container, a painter, one that sizes a side of its own, one\n";
    $out .= "# whose own label is its accessible name), and its keywords in the\n";
    $out .= "# order they are written to the engine.\n";
    $out .= "our \@ELEMENTS = (\n";
    for my $el (@ELEMENTS) {
        my @f = ("name => " . q_str($el->{name}), "kind => $KIND_ID{$el->{name}}", "pix => " . q_str(pix_element($el)));
        push @f, "children => 1" if $el->{children};
        push @f, "paints => 1" if $el->{paints};
        push @f, "native => " . q_str($el->{native}) if defined $el->{native};
        push @f, "owns_label => 1" if $el->{owns_label};
        push @f, "primary => " . q_str($el->{primary}) if defined $el->{primary};
        push @f, "doc => " . q_str($el->{doc}) if defined $el->{doc};
        $out .= "    {\n        " . join(",\n        ", @f) . ",\n        props => [\n";
        for my $p (@{ $el->{props} }) {
            my @g = ("name => " . q_str($p->{name}), "key => $KEY_ID{$p->{name}}");
            if (defined $p->{handler}) {
                push @g, "handler => " . q_str($p->{handler}),
                         "pix => " . q_str(pix_prop($p)),
                         "payload => " . q_str($p->{payload} // $PAYLOAD_NAME{ $p->{handler} });
            } else {
                push @g, "type => " . q_str($p->{type}), "pix => " . q_str(pix_prop($p));
                push @g, "pos => 1" if $p->{pos};
                push @g, "default => " . perl_default($p) if exists $p->{default};
            }
            $out .= "            { " . join(', ', @g) . " },\n";
        }
        $out .= "        ],\n    },\n";
    }
    $out .= ");\nour %ELEMENT = map { \$_->{name} => \$_ } \@ELEMENTS;\n\n";

    $out .= "# The canvas's drawing commands. Not elements: no kind, no handle, no\n";
    $out .= "# keywords that every element takes, and no meaning outside the canvas\n";
    $out .= "# they are written in.\n";
    $out .= "our \@OPS = (\n";
    for my $op (@OPS) {
        my @f = ("name => " . q_str($op->{name}), "pix => " . q_str($op->{pix}),
                 'params => [' . join(', ', map {
                     '{ ' . join(', ', "name => " . q_str($_->{name}), "pix => " . q_str($_->{pix}),
                                       "type => " . q_str($_->{type}),
                                       (exists $_->{default} ? ('default => ' . q_str($_->{default})) : ())) . ' }'
                 } @{ $op->{params} }) . ']');
        $out .= "    { " . join(', ', @f) . " },\n";
    }
    $out .= ");\nour %OP = map { \$_->{name} => \$_ } \@OPS;\n\n1;\n";
    return $out;
}

# --- lib/Rakugan/Elements.pm ------------------------------------------------
# One sub per element. The work is in the runtime, read off the table;
# the subs are the names an app calls, each with its sentence.

sub wrap_doc {
    my ($doc) = @_;
    my @lines;
    my $line = '#';
    for my $w (split ' ', $doc) {
        if (length($line) + length($w) + 1 > 74) { push @lines, $line; $line = '#' }
        $line .= " $w";
    }
    push @lines, $line;
    return join("\n", @lines) . "\n";
}

sub gen_elements {
    my $out = "package Rakugan::Elements;\n" . banner() . <<'HEAD';
#
# The elements an app writes its screen with: one sub per row of the
# table. Positional arguments come first, then keywords and children in
# any order (`column(text(...), row(...), spacing => 8)`); the keywords
# every element takes ride along with the element's own. What each one
# does with what it was given is in Rakugan::Runtime, read off the
# table, so the subs here are the names and their sentences.
use v5.40;
use Rakugan::Runtime;
use Rakugan::Vocab;

HEAD
    $out .= "our \@ELEMENTS = qw(" . join(' ', map { $_->{name} } @ELEMENTS) . ");\n";
    for my $el (@ELEMENTS) {
        $out .= "\n";
        $out .= wrap_doc($el->{doc}) if defined $el->{doc};
        $out .= "sub $el->{name} { Rakugan::Runtime::element(\$Rakugan::Vocab::ELEMENT{$el->{name}}, \@_) }\n";
    }
    $out .= "\n";
    $out .= <<'PAINT';
# --- the canvas's drawing commands ------------------------------------------
#
# A command joins the canvas that is open where it stands. A canvas
# cannot hold another, so one place to remember the open one is enough.

our @OPS = qw(PAINTNAMES);

PAINT
    $out =~ s/PAINTNAMES/join ' ', map { $_->{name} } @OPS/e;
    for my $op (@OPS) {
        $out .= '# ' . join(', ', map { exists $_->{default} ? "$_->{name} => …" : $_->{name} }
                                  @{ $op->{params} }) . ".\n";
        $out .= "sub $op->{name} { Rakugan::Runtime::paint_op('$op->{name}', \@_) }\n";
    }
    $out .= "\n1;\n";
    return $out;
}

# --- write, or check --------------------------------------------------------

my %FILES = (
    "$ROOT/lib/Rakugan/Vocab.pm"    => gen_vocab(),
    "$ROOT/lib/Rakugan/Elements.pm" => gen_elements(),
);

my $check = grep { $_ eq '--check' } @ARGV;
my @stale;
for my $path (sort keys %FILES) {
    my $body = $FILES{$path};
    if ($check) {
        my $on_disk = -f $path ? do { open my $fh, '<:encoding(UTF-8)', $path or die; local $/; <$fh> } : undef;
        push @stale, $path unless defined $on_disk && $on_disk eq $body;
    } else {
        open my $fh, '>:encoding(UTF-8)', $path or die "$path: $!\n";
        print {$fh} $body;
        close $fh;
        (my $rel = $path) =~ s{^\Q$REPO\E/}{};
        print "wrote $rel\n";
    }
}
if ($check && @stale) {
    warn "gen: these are not what elements.toml says — run rakugan/tools/gen.pl\n";
    for my $path (@stale) {
        (my $rel = $path) =~ s{^\Q$REPO\E/}{};
        warn "  $rel\n";
    }
    exit 1;
}
